sklearn/metrics/_regression.py

```diff
@@ -896,3 +896,106 @@ def mean_gamma_deviance(y_true, y_pred, *, sample_weight=None):
+    predicts the expected value of y, disregarding the input features,
+    would get a D^2 score of 0.0.
+
+    Read more in the :ref:`User Guide <d2_tweedie_score>`.
+
+    Parameters
+    ----------
+    y_true : array-like of shape (n_samples,)
+        Ground truth (correct) target values.
+
+    y_pred : array-like of shape (n_samples,)
+        Estimated target values.
+
+    sample_weight : array-like of shape (n_samples,), optional
+        Sample weights.
+
+    power : float, default=0
+        Tweedie power parameter. Either power <= 0 or power >= 1.
+
+        The higher `p` the less weight is given to extreme
+        deviations between true and predicted targets.
+
+        - power < 0: Extreme stable distribution. Requires: y_pred > 0.
+        - power = 0 : Normal distribution, output corresponds to
+          mean_squared_error. y_true and y_pred can be any real numbers.
+        - power = 1 : Poisson distribution. Requires: y_true >= 0 and
+          y_pred > 0.
+        - 1 < p < 2 : Compound Poisson distribution. Requires: y_true >= 0
+          and y_pred > 0.
+        - power = 2 : Gamma distribution. Requires: y_true > 0 and y_pred > 0.
+        - power = 3 : Inverse Gaussian distribution. Requires: y_true > 0
+          and y_pred > 0.
+        - otherwise : Positive stable distribution. Requires: y_true > 0
+          and y_pred > 0.
+
+    Returns
+    -------
+    z : float or ndarray of floats
+        The D^2 score.
+
+    Notes
+    -----
+    This is not a symmetric function.
+
+    Like R^2, D^2 score may be negative (it need not actually be the square of
+    a quantity D).
+
+    This metric is not well-defined for single samples and will return a NaN
+    value if n_samples is less than two.
+
+    References
+    ----------
+    .. [1] Eq. (3.11) of Hastie, Trevor J., Robert Tibshirani and Martin J.
+           Wainwright. "Statistical Learning with Sparsity: The Lasso and
+           Generalizations." (2015). https://trevorhastie.github.io
+
+    Examples
+    --------
+    >>> from sklearn.metrics import d2_tweedie_score
+    >>> y_true = [0.5, 1, 2.5, 7]
+    >>> y_pred = [1, 1, 5, 3.5]
+    >>> d2_tweedie_score(y_true, y_pred)
+    0.285...
+    >>> d2_tweedie_score(y_true, y_pred, power=1)
+    0.487...
+    >>> d2_tweedie_score(y_true, y_pred, power=2)
+    0.630...
+    >>> d2_tweedie_score(y_true, y_true, power=2)
+    1.0
+    """
+    y_type, y_true, y_pred, _ = _check_reg_targets(
+        y_true, y_pred, None, dtype=[np.float64, np.float32])
+    if y_type == 'continuous-multioutput':
+        raise ValueError("Multioutput not supported in d2_tweedie_score")
+    check_consistent_length(y_true, y_pred, sample_weight)
+
+    # TODO: Do we need this?
+    if _num_samples(y_pred) < 2:
+        msg = "D^2 score is not well-defined with less than two samples."
+        warnings.warn(msg, UndefinedMetricWarning)
+        return float('nan')
```

Comment:
This is the exact same behavior as for `r2_score`. I think there is nothing wrong with a single sample. So I'll remove that warning.

Reply:
But single sample does not make sense, so ValueError it is.

Reply:
Should we also chance from warning to raising ValueError for r2_score?

Reply:
Raising `ValueError`, however, produces failure in `test_single_sample`/`check_single_sample` which has the following comment:
```
# Non-regression test: scores should work with a single sample.
# This is important for leave-one-out cross validation.
```
So I revert to returning `nan` for both `r2_score` and `d2_score`.

Reply:
Hum... Ok, but then we have a design issue for LOO, no?

Reply:
I'm not sure. `r2_score` is defined for one sample, it just doesn't make sense. Therefore, we'd like to either warn or raise ValueError. As to LOO, one should use MSE instead of R2. Same applies for D2 and tweedie deviances.

Reply:
I can also change `test_single_sample` and let both R2 and D2 raise ValueError. In either case, a note somewhere in the docstring or user guide concerning LOO would be good.

Reply:
So, what should we do:
1. warn (and keep behavior of `r2_score`)
2. raise `ValueError`